detectors/aws/aws_test.go

```diff
@@ -0,0 +1,118 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/semconv"
+)
+
+func TestAWS_Detect(t *testing.T) {
+	type fields struct {
+		Client client
+	}
+
+	type want struct {
+		Error    string
+		Resource *resource.Resource
+	}
+
+	testTable := map[string]struct {
+		Fields fields
+		Want   want
+	}{
+		"Unavailable": {
+			Fields: fields{Client: &clientMock{}},
+		},
+		"Instance ID Error": {
+			Fields: fields{
+				Client: &clientMock{available: true, idDoc: func() (ec2metadata.EC2InstanceIdentityDocument, error) {
+					return ec2metadata.EC2InstanceIdentityDocument{}, errors.New("id not available")
+				}},
+			},
+			Want: want{Error: "id not available"},
+		},
+		"Instance ID Available": {
+			Fields: fields{
+				Client: &clientMock{available: true, idDoc: func() (ec2metadata.EC2InstanceIdentityDocument, error) {
+					// Example from https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+					doc := ec2metadata.EC2InstanceIdentityDocument{
+						MarketplaceProductCodes: []string{"1abc2defghijklm3nopqrs4tu"},
+						AvailabilityZone:        "us-west-2b",
+						PrivateIP:               "10.158.112.84",
+						Version:                 "2017-09-30",
+						Region:                  "us-west-2",
+						InstanceID:              "i-1234567890abcdef0",
+						InstanceType:            "t2.micro",
+						AccountID:               "123456789012",
+						PendingTime:             time.Date(2016, time.November, 19, 16, 32, 11, 0, time.UTC),
+						ImageID:                 "ami-5fb8c835",
+						Architecture:            "x86_64",
+					}
+
+					return doc, nil
+				}},
+			},
+			Want: want{Resource: resource.New(
+				semconv.CloudProviderAWS,
```

Comment:
I noticed while making this change that there are several other `semconv` values that could be populated by AWS instance metadata, based on what is included in the example identity doc:

* `host.image.id` -> `ami-5fb8c835`
* `host.type` -> `t2.micro`
* `cloud.zone` -> `us-west-2b`

Reply:
@dackroyd can you open an issue to track this? These seem like good things for this detector to include.

Also, I can open an issue if you'd prefer.

Reply:
Sure, I'll raise another issue for that one. Pretty simple change, I can probably push a PR for it as well, but might need confirmation that there isn't anything else that should also be captured before doing that work.

I've not worked in a GCP environment, so I'm not sure what else might be available to add to that detector.

Reply:
Created #403